src/libutil/fs-sink.hh

```diff
@@ -9,27 +9,38 @@
 namespace nix {
 
 /**
- * \todo Fix this API, it sucks.
+ * Actions on an open regular file in the process of creating it.
+ *
+ * See `ParseSink::createRegularFile`.
  */
-struct ParseSink
+struct CreateRegularFile : Sink
 {
-    virtual void createDirectory(const Path & path) = 0;
-
-    virtual void createRegularFile(const Path & path) = 0;
-    virtual void receiveContents(std::string_view data) = 0;
     virtual void isExecutable() = 0;
```

Comment:
Should `isExecutable` be called before, during or after the sink operations?

Reply:
I think any order is fine, actually.

Reply:
What if a NAR has multiple `"executable"` strings per file? That'd not be canonical. It doesn't quite smell right, but I don't have a suggestion for this PR.